vscode/src/chatAgent.ts

```diff
@@ -0,0 +1,214 @@
+    request: vscode.ChatRequest,
+    context: vscode.ChatContext,
+    stream: vscode.ChatResponseStream,
+    token: vscode.CancellationToken,
+  ) {
+    if (this.withinConversation("design", request, context)) {
+      return this.runDesignCommand(request, context, stream, token);
+    }
+
+    stream.markdown(
+      "Please indicate which command you would like to use for our chat.",
+    );
+    return { metadata: { command: "" } };
+  }
+
+  // Logic for the domain driven design command
+  private async runDesignCommand(
+    request: vscode.ChatRequest,
+    context: vscode.ChatContext,
+    stream: vscode.ChatResponseStream,
+    token: vscode.CancellationToken,
+  ) {
+    const previousInteractions = this.previousInteractions(context);
+    const messages = [
+      vscode.LanguageModelChatMessage.User(`User prompt: ${request.prompt}`),
+      vscode.LanguageModelChatMessage.User(DESIGN_PROMPT),
+      vscode.LanguageModelChatMessage.User(
+        `Previous interactions with the user: ${previousInteractions}`,
+      ),
+    ];
+    const workspace = await this.showWorkspacePick();
+
+    // On the first interaction with the design command, we gather the application's schema and include it as part of
+    // the prompt
+    if (request.command && workspace) {
+      const schema = await this.schema(workspace);
+
+      if (schema) {
+        messages.push(
+          vscode.LanguageModelChatMessage.User(
+            `Existing application schema: ${schema}`,
+          ),
+        );
+      }
+    }
+
+    try {
+      // Select the LLM model
+      const [model] = await vscode.lm.selectChatModels({
+        vendor: "copilot",
+        family: "gpt-4-turbo",
+      });
+
+      stream.progress("Designing the models for the requested concept...");
+      const chatResponse = await model.sendRequest(messages, {}, token);
+
+      let response = "";
+      for await (const fragment of chatResponse.text) {
+        // Maybe show the buttons here and display multiple shell blocks?
+        stream.markdown(fragment);
+        response += fragment;
+      }
+
+      const match = /(?<=```shell)[^.$]*(?=```)/.exec(response);
+
+      if (workspace && match && match[0]) {
+        // The shell code block includes all of the `rails generate` commands. We need to strip out the `rails generate`
+        // from all of them since our commands only accept from the generator forward
+        const commandList = match[0]
+          .trim()
+          .split("\n")
+          .map((command) => {
+            return command.replace(/\s*(bin\/rails|rails) generate\s*/, "");
+          });
+
+        stream.button({
+          command: Command.RailsGenerate,
+          title: "Generate with Rails",
+          arguments: [commandList, workspace],
+        });
+
+        stream.button({
+          command: Command.RailsDestroy,
+          title: "Revert previous generation",
+          arguments: [commandList, workspace],
+        });
+      }
```

Comment:
We ask in the prompt to return Rails generate commands inside a shell Markdown code block. The reason is so that we can extract which commands were proposed right here and display the buttons for Rails generate and destroy.